File: server/php/src/Application/Domain/Payments/PaymentDetails/DccProposal.php
<?php

namespace OnlinePayments\ExampleApp\Application\Domain\Payments\PaymentDetails;

final class DccProposal
{
    public ?AmountOfMoney $baseAmount = null;
    public ?string $disclaimerDisplay = null;
    public ?string $disclaimerReceipt = null;
    public ?RateDetails $rate = null;
    public ?AmountOfMoney $targetAmount = null;

    public function toArray(): array
    {
        return [
            'baseAmount' => ($this->baseAmount ?? new AmountOfMoney())->toArray(),
            'disclaimerDisplay' => $this->disclaimerDisplay,
            'disclaimerReceipt' => $this->disclaimerReceipt,
            'rate' => ($this->rate ?? new RateDetails())->toArray(),
            'targetAmount' => ($this->targetAmount ?? new AmountOfMoney())->toArray(),
        ];
    }
}

File: server/ruby/app/business/domain/payments/payment_details/dcc_proposal.rb
module Business
  module Domain
    module Payments
      module PaymentDetails
        class DccProposal
          attr_accessor :base_amount, :disclaimer_display, :disclaimer_receipt, :rate, :target_amount

          def initialize(base_amount: nil, disclaimer_display: nil, disclaimer_receipt: nil, rate: nil, target_amount: nil)
            @base_amount = base_amount
            @disclaimer_display = disclaimer_display
            @disclaimer_receipt = disclaimer_receipt
            @rate = rate
            @target_amount = target_amount
          end

          def to_h
            {
              'baseAmount' => @base_amount&.to_h,
              'disclaimerDisplay' => @disclaimer_display,
              'disclaimerReceipt' => @disclaimer_receipt,
              'rate' => @rate&.to_h,
              'targetAmount' => @target_amount&.to_h
            }
          end
        end
      end
    end
  end
end
